File: core/src/main/kotlin/com/backend/core/auth/AuthController.kt
package com.backend.core.auth

import com.backend.core.base.BaseController
import com.backend.core.security.JwtService
import com.backend.core.security.TokenBlacklistService
import com.backend.core.user.UserEntity
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*

@RequestMapping("/auth")
class AuthController(
    private val authService: AuthService,
    private val tokenBlacklistService: TokenBlacklistService,
    private val jwtService: JwtService
) : BaseController() {

    @PostMapping("/register")
    fun register(@RequestBody request: Map<String, String>): ResponseEntity<Map<String, Any>> {
        return try {
            val email = request["email"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "Email required"))
            val password = request["password"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "Password required"))
            val passwordConfirm = request["passwordConfirm"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "Password confirmation required"))
            val firstName = request["firstName"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "First name required"))
            val lastName = request["lastName"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "Last name required"))

            val result = authService.register(email, password, passwordConfirm, firstName, lastName)
            ResponseEntity.status(HttpStatus.CREATED).body(result)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to (e.message ?: "Registration failed")))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(mapOf("error" to "Registration failed"))
        }
    }

    @PostMapping("/login")
    fun login(@RequestBody request: Map<String, String>): ResponseEntity<Any> {
        return try {
            val email = request["email"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "Email required"))
            val password = request["password"] ?: return ResponseEntity.badRequest().body(mapOf("error" to "Password required"))
            val tenantId = request["tenantId"]

            val tokenPair = authService.login(email, password, tenantId)

            val response = mapOf(
                "accessToken" to tokenPair.accessToken,
                "refreshToken" to tokenPair.refreshToken,
                "tokenType" to tokenPair.tokenType,
                "expiresIn" to tokenPair.expiresIn
            )

            ResponseEntity.ok(response)
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(mapOf("error" to "Invalid email or password"))
        }
    }

    @PostMapping("/refresh")
    fun refreshToken(@RequestBody request: Map<String, String>): ResponseEntity<Any> {
        return try {
            val refreshToken = request["refreshToken"]
                ?: return ResponseEntity.badRequest().body(mapOf("error" to "Refresh token required"))

            val tokenPair = authService.refreshToken(refreshToken)

            val response = mapOf(
                "accessToken" to tokenPair.accessToken,
                "refreshToken" to tokenPair.refreshToken,
                "tokenType" to tokenPair.tokenType,
                "expiresIn" to tokenPair.expiresIn
            )

            ResponseEntity.ok(response)
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(mapOf("error" to "Failed to refresh token"))
        }
    }

    @PostMapping("/logout")
    fun logout(
        @RequestHeader("Authorization") authorization: String
    ): ResponseEntity<Map<String, String>> {
        val token = authorization.removePrefix("Bearer ").trim()

        val expirationMillis = jwtService.getExpiration(token).time - System.currentTimeMillis()
        val expiresInSeconds = expirationMillis / 1000

        tokenBlacklistService.blacklistToken(token, expiresInSeconds)

        return ResponseEntity.ok(mapOf("message" to "Logged out successfully"))
    }


    @GetMapping("/me")
    fun getCurrentUser(@AuthenticationPrincipal user: UserEntity): ResponseEntity<UserEntity> {
        return try {
            val currentUser = authService.getCurrentUser(user.id)
            ResponseEntity.ok(currentUser)
        } catch (e: Exception) {
            ResponseEntity.badRequest().build()
        }
    }

    @PostMapping("/change-password")
    fun changePassword(
        @AuthenticationPrincipal user: UserEntity,
        @RequestBody request: Map<String, String>
    ): ResponseEntity<Map<String, String>> {
        return try {
            val currentPassword = request["currentPassword"]
                ?: return ResponseEntity.badRequest().body(mapOf("error" to "Current password required"))
            val newPassword = request["newPassword"]
                ?: return ResponseEntity.badRequest().body(mapOf("error" to "New password required"))
            val newPasswordConfirm = request["newPasswordConfirm"]
                ?: return ResponseEntity.badRequest().body(mapOf("error" to "Password confirmation required"))

            authService.changePassword(user.id, currentPassword, newPassword, newPasswordConfirm)
            ResponseEntity.ok(mapOf("message" to "Password changed successfully"))
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to (e.message ?: "Failed to change password")))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(mapOf("error" to "Failed to change password"))
        }
    }

    @GetMapping("/check-email/{email}")
    fun checkEmailAvailability(@PathVariable email: String): ResponseEntity<Map<String, Any>> {
        return try {
            val exists = authService.userRepository.existsByEmail(email)
            ResponseEntity.ok(mapOf(
                "email" to email,
                "available" to !exists
            ))
        } catch (e: Exception) {
            ResponseEntity.badRequest().build()
        }
    }

    @GetMapping("/verify-email")
    fun verifyEmail(@RequestParam token: String): ResponseEntity<Map<String, Any>> {
        return try {
            val result = authService.verifyEmail(token)
            ResponseEntity.ok(result)
        } catch (e: IllegalArgumentException) {
            ResponseEntity.badRequest().body(mapOf("error" to (e.message ?: "Verification failed")))
        } catch (e: Exception) {
            ResponseEntity.badRequest().body(mapOf("error" to "Verification failed"))
        }
    }
}

File: identity-service/src/auth/auth.controller.ts
import {
  BadRequestException,
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Post,
  Request,
  UnauthorizedException,
  UseGuards,
} from '@nestjs/common';
import { AuthService } from './auth.service';
import { Public } from './decorators/public.decorator';
import { JwtAuthGuard } from './guards/jwt-auth.guard';
import {
  AccountLockedException,
  AuthUser,
  InvalidCredentialsException,
  LoginDto,
  LoginResponse,
  RefreshTokenDto,
  RegisterDto,
  TwoFactorRequiredException,
} from './types';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Public()
  @Post('login')
  @HttpCode(HttpStatus.OK)
  async login(@Body() loginDto: LoginDto): Promise<LoginResponse> {
    try {
      const tokenPair = await this.authService.login(loginDto);

      return {
        success: true,
        data: tokenPair,
      };
    } catch (error) {
      if (error instanceof TwoFactorRequiredException) {
        return {
          success: false,
          error: '2FA_REQUIRED',
          message: error.message,
        };
      }

      if (
        error instanceof InvalidCredentialsException ||
        error instanceof AccountLockedException
      ) {
        throw new UnauthorizedException(error.message);
      }

      throw new UnauthorizedException('Login failed');
    }
  }

  @Public()
  @Post('register')
  @HttpCode(HttpStatus.CREATED)
  async register(@Body() registerDto: RegisterDto) {
    try {
      const result = await this.authService.register(registerDto);

      return {
        success: true,
        data: result,
      };
    } catch (error) {
      if (error instanceof BadRequestException) {
        throw error;
      }

      throw new BadRequestException('Registration failed');
    }
  }

  @Public()
  @Post('refresh')
  @HttpCode(HttpStatus.OK)
  async refreshToken(@Body() refreshTokenDto: RefreshTokenDto) {
    try {
      const tokenPair = await this.authService.refreshToken(refreshTokenDto);

      return {
        success: true,
        data: tokenPair,
      };
    } catch (error) {
      throw new UnauthorizedException('Failed to refresh token');
    }
  }

  @UseGuards(JwtAuthGuard)
  @Get('me')
  async getCurrentUser(@Request() req: any) {
    const authUser: AuthUser = req.user;

    try {
      const user = await this.authService.getCurrentUser(authUser.id);

      return {
        success: true,
        data: user,
      };
    } catch (error) {
      throw new UnauthorizedException('Failed to get current user');
    }
  }

  @UseGuards(JwtAuthGuard)
  @Post('logout')
  @HttpCode(HttpStatus.OK)
  async logout(@Request() req: any) {
    const authUser: AuthUser = req.user;

    try {
      const result = await this.authService.logout(authUser.id);

      return {
        success: true,
        ...result,
      };
    } catch (error) {
      throw new BadRequestException('Logout failed');
    }
  }

  @UseGuards(JwtAuthGuard)
  @Post('change-password')
  @HttpCode(HttpStatus.OK)
  async changePassword(
    @Request() req: any,
    @Body()
    body: {
      currentPassword: string;
      newPassword: string;
      newPasswordConfirm: string;
    },
  ) {
    const authUser: AuthUser = req.user;
    const { currentPassword, newPassword, newPasswordConfirm } = body;

    try {
      await this.authService.changePassword(
        authUser.id,
        currentPassword,
        newPassword,
        newPasswordConfirm,
      );

      return {
        success: true,
        message: 'Password changed successfully',
      };
    } catch (error) {
      if (error instanceof BadRequestException) {
        throw error;
      }

      throw new BadRequestException('Failed to change password');
    }
  }
}
